File: Backend-node/src/controllers/taskController.js
const createTask = async (req, res) => {
    try {
        const { title, description, user_id, due_date, priority, status } = req.body;

        if (!title || !description || !user_id || !due_date || !priority || !status) {
            return res.status(400).json({
                message: "Todos los campos son obligatorios: title, description, user_id, due_date, priority, status.",
                status: false,
            });
        }

        const response = await fetch(`${process.env.APIGATEWAY}/Test/create-task`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                body: JSON.stringify({
                    title,
                    description,
                    user_id,
                    due_date,
                    priority,
                    status,
                })
            }),
        });

        const data = await response.json();

        if (!response.ok) {
            throw new Error(data.message || 'Error creating task');
            
        }

        // Deserializamos el cuerpo que viene en 'data.body'
        const responseData = JSON.parse(data.body);

        // Devolvemos una respuesta exitosa con `status: true`
        return res.status(response.status).json({
            message: responseData.message,
            taskId: responseData.taskId,
            status: responseData.status,
        });
    } catch (error) {
        console.error('Error creating task:', error);

        // En caso de error, devolvemos `status: false`
        return res.status(500).json({
            message: 'Error interno al crear la tarea',
            error: error.message,
            status: false,
        });
    }
};

const updateTask = async (req, res) => {
    const { task_id } = req.params;
    try {
        const { title, description, due_date, priority, status } = req.body;
        if (!task_id ) {
            return res.status(400).json({
                message: "Todos los campos son obligatorios: task_id.",
                status: false,
            });
        }

        const response = await fetch(`${process.env.APIGATEWAY}/Test/updateTask`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                body: JSON.stringify({
                    title,
                    description,
                    task_id,
                    due_date,
                    priority,
                    status,
                })
            }),
        });

        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || 'Error updating task');
        }

        // Deserializamos el cuerpo que viene en 'data.body'
        const responseData = JSON.parse(data.body);

        // Devolvemos una respuesta exitosa con `status: true`
        return res.status(response.status).json({
            message: responseData.message,
            taskId: responseData.taskId,
            status: responseData.status,
        });
    } catch (error) {
        console.error('Error updating task:', error);

        // En caso de error, devolvemos `status: false`
        return res.status(500).json({
            message: 'Error interno al actualizar la tarea',
            error: error.message,
            status: false,
        });
    }
};

const deleteTask = async (req, res) => {
    const { task_id } = req.params;
    try {
        if (!task_id ) {
            return res.status(400).json({
                message: "Todos los campos son obligatorios: task_id.",
                status: false,
            });
        }

        const response = await fetch(`${process.env.APIGATEWAY}/Test/deleteTask`, {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                body: JSON.stringify({
                    task_id
                })
            }),
        });

        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || 'Error delete task');
        }

        // Deserializamos el cuerpo que viene en 'data.body'
        const responseData = JSON.parse(data.body);

        // Devolvemos una respuesta exitosa con `status: true`
        return res.status(response.status).json({
            message: responseData.message,
            taskId: responseData.taskId,
            status: responseData.status,
        });
    } catch (error) {
        console.error('Error delete task:', error);

        // En caso de error, devolvemos `status: false`
        return res.status(500).json({
            message: 'Error interno al eliminar la tarea',
            error: error.message,
            status: false,
        });
    }
};

const getTask = async (req, res) => {
    const { task_id } = req.params;
    try {
        if (!task_id ) {
            return res.status(400).json({
                message: "Todos los campos son obligatorios: task_id.",
                status: false,
            });
        }

        const response = await fetch(`${process.env.APIGATEWAY}/Test/getTaskById`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                body: JSON.stringify({
                    task_id
                })
            }),
        });

        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || 'Error get task');
        }

        // Deserializamos el cuerpo que viene en 'data.body'
        const responseData = JSON.parse(data.body);

        // Devolvemos una respuesta exitosa con `status: true`
        return res.status(response.status).json({
            task: responseData.task,
            status: responseData.status,
        });
    } catch (error) {
        console.error('Error get task:', error);

        // En caso de error, devolvemos `status: false`
        return res.status(500).json({
            message: 'Error interno al obtener la tarea',
            error: error.message,
            status: false,
        });
    }
};

const getAllTask = async (req, res) => {
    try {

        const response = await fetch(`${process.env.APIGATEWAY}/Test/getAllTask`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            },
        });

        const data = await response.json();
        if (!response.ok) {
            throw new Error(data.message || 'Error get task');
        }

        // Deserializamos el cuerpo que viene en 'data.body'
        const responseData = JSON.parse(data.body);

        // Devolvemos una respuesta exitosa con `status: true`
        return res.status(response.status).json({
            tasks: responseData.tasks,
            status: responseData.status,
        });
    } catch (error) {
        console.error('Error get task:', error);

        // En caso de error, devolvemos `status: false`
        return res.status(500).json({
            message: 'Error interno al obtener la tarea',
            error: error.message,
            status: false,
        });
    }
};

module.exports = {
    createTask,
    updateTask,
    deleteTask,
    getTask,
    getAllTask
};

File: Backend-python/src/controllers/taskController.py
import os
from flask import request, jsonify


def createTask():
    try:
        data = request.get_json()
        title = data.get('title')
        description = data.get('description')
        user_id = data.get('user_id')
        due_date = data.get('due_date')
        priority = data.get('priority')
        status = data.get('status')

        if not title or not description or not user_id or not due_date or not priority or not status:
            return jsonify({
                'message': "Todos los campos son obligatorios: title, description, user_id, due_date, priority, status.",
                'status': False,
            }), 400
        

        body = {
            'title': title,
            'description': description,
            'user_id': user_id,
            'due_date': due_date,
            'priority': priority,
            'status': status,
        }

        response = request.post(
            f"{os.getenv('APIGATEWAY')}/Test/create-task",
            json={'body': body}
        )

        if not response.ok:
            data = response.json()
            raise Exception(data.get('message', 'Error creating task'))

        
        response_data = response.json()
        task_response = response_data.get('body')

        return jsonify({
            'message': task_response['message'],
            'taskId': task_response['taskId'],
            'status': task_response['status'],
        }), response_data['status']
    
    except Exception as error:
        print('Error creating task:', error)

        return jsonify({
            'message': 'Error interno al crear la tarea',
            'error': str(error),
            'status': False,
        }), 500


def updateTask(task_id):
    try:
        data = request.get_json()
        title = data.get('title')
        description = data.get('description')
        due_date = data.get('due_date')
        priority = data.get('priority')
        status = data.get('status')

        if not task_id:
            return jsonify({
                'message': "El campo task_id es obligatorio.",
                'status': False,
            }), 400

        body = {
            'title': title,
            'description': description,
            'task_id': task_id,
            'due_date': due_date,
            'priority': priority,
            'status': status,
        }

        response = request.put(
            f"{os.getenv('APIGATEWAY')}/Test/updateTask",
            json={'body': body}
        )

        if not response.ok:
            data = response.json()
            raise Exception(data.get('message', 'Error updating task'))

        response_data = response.json()
        task_response = response_data.get('body')

        return jsonify({
            'message': task_response['message'],
            'taskId': task_response['taskId'],
            'status': task_response['status'],
        }), response_data['status']

    except Exception as error:
        print('Error updating task:', error)

        return jsonify({
            'message': 'Error interno al actualizar la tarea',
            'error': str(error),
            'status': False,
        }), 500
    
def deleteTask(task_id):
    try:
        if not task_id:
            return jsonify({
                'message': "El campo task_id es obligatorio.",
                'status': False,
            }), 400

        response = request.delete(
            f"{os.getenv('APIGATEWAY')}/Test/deleteTask",
            json={'body': {'task_id': task_id}}  
        )

        if not response.ok:
            data = response.json()
            raise Exception(data.get('message', 'Error deleting task'))

        response_data = response.json()
        task_response = response_data.get('body')

        return jsonify({
            'message': task_response['message'],
            'taskId': task_response['taskId'],
            'status': task_response['status'],
        }), response_data['status']

    except Exception as error:
        print('Error deleting task:', error)

        return jsonify({
            'message': 'Error interno al eliminar la tarea',
            'error': str(error),
            'status': False,
        }), 500


def getTask(task_id):
    try:
        if not task_id:
            return jsonify({
                'message': "El campo task_id es obligatorio.",
                'status': False,
            }), 400

        response = request.post(
            f"{os.getenv('APIGATEWAY')}/Test/getTaskById",
            json={'body': {'task_id': task_id}}  
        )

        if not response.ok:
            data = response.json()
            raise Exception(data.get('message', 'Error get task'))

        response_data = response.json()
        task_response = response_data.get('body')

        return jsonify({
            'task': task_response.get('task'),
            'status': task_response.get('status'),
        }), response_data['status']

    except Exception as error:
        print('Error get task:', error)

        return jsonify({
            'message': 'Error interno al obtener la tarea',
            'error': str(error),
            'status': False,
        }), 500

def getAllTask():
    try:
        response = request.get(
            f"{os.getenv('APIGATEWAY')}/Test/getAllTask",
            headers={'Content-Type': 'application/json'}
        )

        if not response.ok:
            data = response.json()
            raise Exception(data.get('message', 'Error get task'))

        response_data = response.json()
        tasks_response = response_data.get('body')

        return jsonify({
            'tasks': tasks_response.get('tasks'),
            'status': tasks_response.get('status'),
        }), response.status_code

    except Exception as error:
        print('Error get task:', error)

        return jsonify({
            'message': 'Error interno al obtener las tareas',
            'error': str(error),
            'status': False,
        }), 500